cmd/composer-cli/blueprints/diff.go

```diff
@@ -25,7 +31,189 @@ func init() {
+		err := getBlueprintJSON(args[0], args[1])
+		if err != nil {
+			return root.ExecutionError(cmd, err.Error())
+		}
+
+		err = getBlueprintJSON(args[0], args[2])
+		if err != nil {
+			return root.ExecutionError(cmd, err.Error())
+		}
+		return nil
+	}
+
+	fromBlueprint, err := getBlueprint(args[0], args[1])
+	if err != nil {
+		return root.ExecutionError(cmd, err.Error())
+	}
+
+	toBlueprint, err := getBlueprint(args[0], args[2])
+	if err != nil {
+		return root.ExecutionError(cmd, err.Error())
+	}
+
+	// Everything after args[2] is passed to diff
+	// default to --color and -u if nothing passed
+	diffArgs := args[3:]
+	if len(diffArgs) == 0 {
+		diffArgs = []string{"--color", "-u"}
+	}
+
+	err = runDiff(fromBlueprint, args[1], toBlueprint, args[2], diffArgs)
+	if err != nil {
+		return root.ExecutionError(cmd, "%s", err)
+	}
+
+	return nil
+}
+
+func getBlueprint(name, commit string) (string, error) {
+	if commit == "WORKSPACE" {
+		// If nothing has been pushed to the /blueprints/workspace this will be the latest
+		bps, resp, err := root.Client.GetBlueprintsTOML([]string{name})
+		if err != nil {
+			return "", err
+		}
+		if resp != nil && !resp.Status {
+			return "", fmt.Errorf(strings.Join(resp.AllErrors(), ", "))
+		}
+		if len(bps) == 0 {
+			return "", fmt.Errorf("no blueprints")
+		}
+		return bps[0], nil
+	} else if commit == "NEWEST" {
+		// Get the list of changes for this blueprint
+		bps, resp, err := root.Client.GetBlueprintsChanges([]string{name})
+		if err != nil {
+			return "", err
+		}
+		if len(resp) > 0 {
+			return "", fmt.Errorf(resp[0].String())
+		}
+		if len(bps) < 1 {
+			return "", fmt.Errorf("no blueprints")
+		}
+		if len(bps[0].Changes) < 1 {
+			return "", fmt.Errorf("no NEWEST commit for %s", name)
+		}
+		commit = bps[0].Changes[0].Commit
+	}
+
+	// Return the blueprint commit
+	bp, resp, err := root.Client.GetBlueprintChangeTOML(name, commit)
+	if err != nil {
+		return "", err
+	}
+	if resp != nil && !resp.Status {
+		return "", fmt.Errorf(strings.Join(resp.AllErrors(), ", "))
+	}
+	return bp, nil
+}
 
+// runDiff runs the system diff utility from a temporary directory
+func runDiff(fromBlueprint, fromCommit, toBlueprint, toCommit string, diffArgs []string) error {
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	// Make a temporary directory to save the blueprints into
+	// defer cleanup so it doesn't ever get left behind
+	tmpDir, err := os.MkdirTemp("", "bp-diff-*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		return err
+	}
+	defer os.Chdir(cwd) //nolint:errcheck
+
+	// Use saveBlueprint from save.go
+	fromFilename, err := saveBlueprint(fromBlueprint, fromCommit, "")
+	if err != nil {
+		return err
+	}
+	diffArgs = append(diffArgs, fromFilename)
+
+	toFilename, err := saveBlueprint(toBlueprint, toCommit, "")
+	if err != nil {
+		return err
+	}
+	diffArgs = append(diffArgs, toFilename)
+
+	_, err = exec.LookPath("diff")
```

Comment:
A lot of people like `vim-diff`, shall we make this configurable via a non-positional argument?

Reply:
If anything it should be an environmental variable, not a cmdline arg. But there's also the issue of the return codes, vimdiff doesn't document them so they probably don't match diff. Since I'm trying to keep this simple, I think something like this can wait for more user input.